Tweak probing logic

martinthomson opened the pull request:
This changes probing logic so that we send multiple packets (2 for now) in every packet number space on PTO... But only the first forces a probe.

This should avoid the problem we had regarding retransmissions that didn't have space for their payload in a single packet on PTO.  That's something we relied on for the reorder_handshake test.

This change produced some pretty significant instability though.  I see anywhere between 4 and 8 tests failing as a result.  Errors are not all reproducible.  That might be because I added 100 to the starting packet number in the client Initial, who knows.  Here's the list:

    connection::tests::ackrate::ack_rate_persistent_congestion
    connection::tests::cc::cc_cong_avoidance_recovery_period_unchanged
    connection::tests::cc::single_packet_on_recovery
    connection::tests::handshake::reorder_05rtt_with_0rtt
    connection::tests::handshake::reorder_handshake
    connection::tests::recovery::loss_time_past_largest_acked
    connection::tests::recovery::pto_handshake_complete
    connection::tests::resumption::remember_smoothed_rtt

mxinden committed:
refactor(crypto): improve `ResumptionToken` `Debug` impl (#2847)

* refactor(crypto): improve `ResumptionToken` `Debug` impl

Use `hex_snip_middle` to print token.

See similar change in https://github.com/mozilla/neqo/pull/2815.

* test: debug impl

mxinden committed:
docs(transport): extend preprocess_packet docs on State::Closing (#2848)

larseggert committed:
ci: Use 6.11 kernel on GCP (#2849)

larseggert committed:
ci: Use implicit `$HOME` home directory (#2850)

larseggert committed:
feat: Optimize `neqo-qpack` (#2831)

* feat: Optimize `neqo-qpack`

 Similar to #2827

* Address @martinthomson's comments

mxinden committed:
fix(http3): reduce visibility of internal functions (#2802)

Mark functions not used in Firefox as pub(crate). This enables us to
detect unused code and simplifies future refactorings.

martinthomson committed:
Tweak probing logic

This changes probing logic so that we send multiple packets (2 for now)
in every packet number space on PTO... But only the first forces a
probe.

This should avoid the problem we had regarding retransmissions that
didn't have space for their payload in a single packet on PTO.  That's
something we relied on for the reorder_handshake test.

This change produced some pretty significant instability though.  I see
anywhere between 4 and 8 tests failing as a result.  Errors are not all
reproducible.  That might be because I added 100 to the starting packet
number in the client Initial, who knows.  Here's the list:

    connection::tests::ackrate::ack_rate_persistent_congestion
    connection::tests::cc::cc_cong_avoidance_recovery_period_unchanged
    connection::tests::cc::single_packet_on_recovery
    connection::tests::handshake::reorder_05rtt_with_0rtt
    connection::tests::handshake::reorder_handshake
    connection::tests::recovery::loss_time_past_largest_acked
    connection::tests::recovery::pto_handshake_complete
    connection::tests::resumption::remember_smoothed_rtt

martinthomson committed:
Logging, tweaking, fixing

martinthomson committed:
Make progress on refining compatible upgrade

martinthomson committed:
Add a test for interleaving of v1 and v2 packets at the server

martinthomson committed:
Tests for interleaving of versions

mxinden committed:
refactor: unify most exchange_packets impls (#2851)

This commit unifies various `exchange_packets` implementations across the
codebase.

aochagavia committed:
Make initial RTT configurable (#2825)

Co-authored-by: Max Inden <mail@max-inden.de>

mxinden committed:
fix(bin): replace buggy `to_headers` with `FromStr` `impl` (#2833)

larseggert commented:
Do you want me to merge this into the neqo PR for further debugging or leave it here?

larseggert committed:
Turn off random initial pns in two tests that need to parse the CI

larseggert commented:
Two of the failing tests were ones that constructed and parsed a CI manually. I disabled pn randomization for those.

I now see these remaining failures (repeatably on every run):
```
     Summary [   4.511s] 1202 tests run: 1199 passed (1 leaky), 3 failed, 2 skipped
        FAIL [   0.188s] neqo-transport connection::tests::recovery::handshake_ack_pto
        FAIL [   0.187s] neqo-transport connection::tests::recovery::loss_time_past_largest_acked
        FAIL [   0.137s] neqo-transport connection::tests::recovery::pto_handshake_complete
```

martinthomson commented:
> Do you want me to merge this into the neqo PR for further debugging or leave it here?

I should have marked it as a draft.  I want to do some cleaning up first.  There are several problems at play here.

I'm down to those three as well.  It's because I've changed our PTO logic in a few subtle ways:

1. Previously, if we had a PTO on multiple packet number spaces, we were reverting to a single packet of PTO, because the packet count was based on the earliest space that had a PTO.  This patch retains the second probe at the Application Data space, rather than allowing it to be masked off.  This leads to some extra probes in some tests, like the first in that list.  Those need to be accounted for.
2. Previously, if we have multiple packet number spaces get a PTO, the last one overrode the spaces in which we probed.  That meant that we were not forcing probes in earlier spaces.  That seems like it would degrade recovery performance, particularly for handshakes.  We do sort of probe anyway, but that wasn't guaranteed, especially if we are up against CC limits.  We do send packets in higher spaces and we had some problems with that as a result.
3. Previously, we would only probe once for Initial and Handshake.  This change makes it so that we force a probe for Application Data twice, but we will now allow the sending of a second packet (and one that ignores CC limits) in the Initial and Handshake spaces.  That won't force the creation of a packet that includes PING, but the option is opened now, rather than skipping those spaces entirely.  That has a subtle effect, but it allows for some of our test cases to pass where they weren't doing so before.  That is because we now have Initial and Handshake messages that have more CRYPTO frame data than would fit in a single packet.

These are small changes for sure.  However, they do have some knock-on effects.  I also want to make sure that we're deliberate about these.  I think that these are improvement, but I want you and @mxinden to be comfortable with the changes before we push on.

The more substantive bug fix here is that we were not doing compatible version upgrade if the client received an Initial that didn't have a complete CRYPTO frame in it.  The server would switch to Version2 and we would believe that we got a version already, so we would fail to decrypt Initial packets from the server.  That could be the cause of connection failures.  I don't know if we'd see that in telemetry, but it's enough of problem that we should really prioritize fixing it.

I'm a little leery of the fix I have here, because our code assumes that a server is always able to make a decision about the version before sending a CRYPTO frame.  That is not true if the server needs to send a HelloRetryRequest in TLS.  So in addition to the cleanup, I think that I need to work out what the right answer is.

https://datatracker.ietf.org/doc/html/rfc9368#section-2.3 doesn't make the same assumptions we do.  The spec assumes that the client infers from the choice of version in the packet header.  So it's possible we're just a little too aggressive about confirming the server's choice.  I have some ideas that I'll try out this afternoon in addition to cleaning up those tests.

martinthomson commented on neqo-transport/src/crypto.rs:
```diff
@@ -990,7 +991,7 @@ impl CryptoStates {
         let min_pn = if randomize_ci_pn {
             let r = random::<2>();
-            packet::Number::from(r[0] & r[1]) + 1
+            packet::Number::from(r[0] & r[1]) + 256
```
Need to revert this tweak.  It's useful for now, because it removes variability.

martinthomson commented on neqo-transport/src/crypto.rs:
```diff
@@ -1043,6 +1045,7 @@ impl CryptoStates {
                 let next = self.initials[confirmed]
                     .as_mut()
                     .ok_or(Error::VersionNegotiation)?;
+                next.rx.continuation(&prev.rx)?;
```
This is the other major bug we had.  One that will definitely lead to handshake failures.  And I'm fairly sure that this fix is NOT going to work.

If a peer randomizes the initial value in the Initial packet number space, we sort of assumed that they could switch back and forth between multiple versions without consequence.  But our packet number reconstruction code uses the largest value from this code to track what has been seen.

On *sending* we look at the largest acknowledged (from the loss recovery code) to determine how small a packet number we can get away with.  If the starting packet number is 256, we might have received an ACK for that packet, so when we encode packet 257, we encode a single byte truncated packet number (0x01).

That's fine for sending, but on our receive side, if the new packet uses a different version *as it will when there is compatible version negotiation* the ACK-only packet we received will be for Version1 and the new packet will be Version2.  The packet we received will update the Version1 crypto state, not the Version2 crypto state.

To be clear, this change also prevents reuse of packet numbers, which isn't necessary for this space.  Confidentiality isn't really a goal and the two contexts have completely different keys, so we aren't even really abusing the crypto.  Still, it makes sense to check, which is what the continuation function does.

While the server side is good, the problem here is that we could get a Version2 packet before we confirm the versions, but AFTER the peer has received an ACK.  I think that is a pretty distinct problem for the client.

The case here would be:
1. client sends two Initial packets at Version1
2. server receives one packet, ACKs first in Version1, uses packet number 500
3. client receives ACK and updates Version1 crypto state to 500
5. server receives second packet, processes handshake, chooses Version2, sends handshake in Version2 packet
6. packet is lost and both client server sends a probe on PTO
7. client probe reaches the server first and contains an ACK, server updates its view of largest acknowledged
8. server sends its handshake, using Version2 and a truncated packet 502 -> 0xf6 (not 0x01f6)
9. client receives handshake, but assumes that it's packet 246, not 502, so it fails to decrypt

In that case, we'll not have run this code at the point that we need to recover the packet number.  I don't have a good solution in mind for that yet.  I'll need to write a test case.

That's made harder, because we're not doing randomizing the starting packet number for servers.  @larseggert, would it be terrible if we randomized the starting packet number for the server, but turned it off by default in tests?

larseggert replied:
> @larseggert, would it be terrible if we randomized the starting packet number for the server, but turned it off by default in tests?

No, we can turn that on, and I can see how much work it would be to fix the tests or otherwise disable the randomization.

martinthomson commented on neqo-transport/src/crypto.rs:
```diff
@@ -1010,6 +1011,7 @@ impl CryptoStates {
                     "[{self}] Continue packet numbers for initial after retry (write is {:?})",
                     prev.rx.used_pn,
                 );
+                initial.rx.continuation(&prev.rx)?;
```
This one is similar to the situation below, except that this code is only really used in a context that might matter at a server.  The situation with servers is far less problematic because servers are the first to commit to a version and clients can only commit to a version after the server has done so.  This is still necessary, but I'm completely confident in this path.

martinthomson committed:
Fix the PTO tests

martinthomson commented:
Step 1 complete: the tests all now pass.

martinthomson committed:
Tests for decryption of compatible upgrade

martinthomson commented:
Step 2 complete: write tests for compatible upgrade and successful decryption.  Next to make those tests fail by having the server randomize the initial packet number too.

martinthomson committed:
Better normalize use of connection parameters

martinthomson committed:
Randomize server Initial packet number as well, rename config param

martinthomson committed:
Failing a test now for client decrypting the v2 Initial

mxinden committed:
refactor(webtransport): rename WebTransportEvent::Session to NewSession (#2853)

On a new WebTransport session a client emits
`neqo_http3::client_events::WebTransportEvent::Session` and a server
emits  `neqo_http3::server_events::WebTransportServerEvent::NewSession`.
For the sake of consistency, rename the former to `NewSession`. Also
consistent with `neqo_http3:client_event::WebTransportEvent::NewStream`.

martinthomson committed:
Checkpoint

martinthomson commented:
OK, all the pieces are in place, except the fix for the server Initial shifting version when the initial packet number is large.

larseggert commented on neqo-transport/src/connection/tests/mod.rs:
```diff
@@ -100,6 +100,10 @@ impl ConnectionIdGenerator for CountingConnectionIdGenerator {
 // These are a direct copy of those functions.
+pub fn client_default_params() -> ConnectionParameters {
+    ConnectionParameters::default()
+}
```
Do we need this, since it's just `ConnectionParameters::default()`? Do you expect this to change?

It also makes the diff super noisy. If we want this, maybe this would be better in a separate `chore` PR.

larseggert commented on neqo-transport/src/connection/tests/mod.rs:
```diff
@@ -128,12 +132,16 @@ fn zero_len_cid_client(local_addr: SocketAddr, remote_addr: SocketAddr) -> Conne
     .unwrap()
 }
 
+pub fn server_default_params() -> ConnectionParameters {
+    ConnectionParameters::default().randomize_first_pn(false)
+}
```
Do we still want the server to also randomize? If so, should we (also) remove this and just do `ConnectionParameters::default()`?

It also makes the diff super noisy. If we want this, maybe this would be better in a separate `chore` PR.

martinthomson replied:
I'll give the randomization on the server side a try.  Maybe we'll get lucky and it won't hurt TOO much.

martinthomson replied:
Not TOO bad, though some of these are not ones that we were previously disabling the client-side randomization.

    connection::tests::handshake::coalesce_05rtt
    connection::tests::handshake::drop_handshake_packet_from_wrong_address
    connection::tests::handshake::dup_server_flight1
    connection::tests::handshake::extra_initial_hs
    connection::tests::handshake::only_server_initial
    connection::tests::keys::discarded_initial_keys
    connection::tests::resumption::ticket_rtt_larger_than_default
    connection::tests::resumption::ticket_rtt_less_than_default
    connection::tests::vn::compatible_upgrade_large_initial

larseggert commented on test-fixture/src/lib.rs:
```diff
@@ -202,25 +214,30 @@ pub fn new_client(params: ConnectionParameters) -> Connection {
+#[must_use]
+pub fn server_default_params_h3() -> ConnectionParameters {
+    server_default_params().pacing(false)
 }
```
I kinda prefer the explicit override of `default()` where we need it, it makes tests that are special more visible. YMMV.

larseggert commented on neqo-transport/src/connection/tests/handshake.rs:
```diff
@@ -1145,7 +1257,7 @@ fn only_server_initial() {
     assert_eq!(client.stats().frame_tx.ack, 0);
     let probe = client.process(Some(server_initial1), now).dgram();
-    assertions::assert_initial(&probe.unwrap(), false);
+    assert_initial(&probe.unwrap(), false);
```
Gratuitous change?

larseggert commented on neqo-transport/src/connection/tests/mod.rs:
```diff
@@ -178,7 +186,18 @@ pub fn rttvar_after_n_updates(n: usize, rtt: Duration) -> Duration {
+    pub fn new() -> Option<Box<dyn test_internal::FrameWriter>> {
+        Some(Box::new(Self {}))
+    }
+}
```
Gratuitous change? Maybe better in a separate PR?

larseggert commented on neqo-transport/src/connection/mod.rs:
```diff
@@ -387,7 +388,7 @@ impl Connection {
     ) -> Res<Self> {
         let dcid = ConnectionId::generate_initial();
-        let randomize_ci_pn = conn_params.randomize_ci_pn_enabled();
+        let randomize_first_pn = conn_params.randomize_first_pn_enabled();
```
Like the name change.

larseggert commented on neqo-transport/src/connection/mod.rs:
```diff
@@ -3062,42 +3059,53 @@ impl Connection {
+            qdebug!("[{self}] Compatible upgrade {:?} ==> {v:?}", self.version);
+            self.version = v;
         }
+        self.crypto.confirm_version(v)?;
         Ok(())
     }
 
     fn handshake(
         &mut self,
         now: Instant,
-        packet_version: Version,
         space: PacketNumberSpace,
+        packet_version: Version,
```
Why flip the argument order?

larseggert commented on neqo-transport/src/connection/mod.rs:
```diff
@@ -2502,6 +2492,13 @@ impl Connection {
+
+            #[cfg(test)]
+            if let Some(w) = &mut self.test_frame_writer {
+                if !builder.is_full() {
+                    w.write_frames(builder);
+                }
```
Should we `debug_assert!` here that the builder is not full? To catch surprises?

larseggert commented on neqo-transport/src/connection/mod.rs:
```diff
@@ -3062,42 +3059,53 @@ impl Connection {
                 .ok_or(Error::ProtocolViolation)?;
-            self.crypto.states_mut().init_server(version, dcid)?;
-            self.confirm_version(version)?;
+            // No need to randomize the starting packet number; that's already taken care of.
```
```suggestion
```

larseggert commented on neqo-transport/src/connection/mod.rs:
```diff
@@ -3206,18 +3216,28 @@ impl Connection {
+                    // This is an assumption, but very often a good one.
+                    // This function checks the state and does nothing if we have a version.
+                    self.compatible_upgrade(packet_version, now)?;
+                } */
```
Did you mean to do something with this or can it be removed?

larseggert commented on neqo-transport/src/packet/mod.rs:
```diff
@@ -803,15 +803,19 @@ impl<'a> Public<'a> {
     fn decrypt_header(&mut self, crypto: &CryptoDxState) -> Res<(bool, Number, Range<usize>)> {
-        assert_ne!(self.packet_type, Type::Retry);
-        assert_ne!(self.packet_type, Type::VersionNegotiation);
+        debug_assert_ne!(self.packet_type, Type::Retry);
+        debug_assert_ne!(self.packet_type, Type::VersionNegotiation);
```
Gratuitous change?

larseggert commented on neqo-transport/src/recovery/mod.rs:
```diff
@@ -71,7 +71,7 @@ impl SendProfile {
             limit: max(ACK_ONLY_SIZE_LIMIT - 1, limit),
             pto: None,
-            probe: PacketNumberSpaceSet::default(),
+            probe: PacketNumberSpaceSet::empty(),
```
Gratuitous change?

larseggert commented on neqo-transport/src/recovery/mod.rs:
```diff
@@ -425,35 +424,22 @@ struct PtoState {
             space,
             count: 1,
-            packets: Self::pto_packet_count(space),
+            packets: MAX_PTO_PACKET_COUNT,
```
Do we still need the `packets` field? It seems to now be always `MAX_PTO_PACKET_COUNT`.

larseggert commented on neqo-transport/src/crypto.rs:
```diff
@@ -988,9 +994,15 @@ impl CryptoStates {
+            // Then maybe add up to 256 to occasionally nudge into a second byte
+            // for both varint and packet number encodings.
+            // And a small offset to ensure that the value is always non-zero.
+            // TODO: change the 300 offset to be 1.
+            (pn(r[0]) & 0x1f) + (pn(r[1]) + pn(r[2]) + pn(r[3])).saturating_sub(512) + 300
```
Should the `TODO` be addressed?

larseggert commented on neqo-transport/tests/conn_vectors.rs:
```diff
@@ -252,10 +250,10 @@ const INITIAL_PACKET_29: &[u8] = &[
     Connection::new_server(
-        test_fixture::DEFAULT_KEYS,
-        test_fixture::DEFAULT_ALPN,
+        DEFAULT_KEYS,
+        DEFAULT_ALPN,
```
Gratuitous change?

larseggert reviewed (commented):
The PR is very noisy, since it includes a bunch of "chore"-style changes that – if we want to make them – should probably be factored out.

mxinden commented:
Thanks for the ping.

If I understand correctly, this pull request addresses the following issues:

- send 2 packets in initial and handshake space, if needed, don't force the second, would fix issue where CRYPTO frame spans two packets (see comment https://github.com/mozilla/neqo/pull/2499#issuecomment-3186237386)
- fully isolate packet count per space, e.g. two probes in application space even when we simultaneously probe the initial space
- probe in all spaces, not just the largest one
- client / server version mismatch on incomplete CRYPTO frame

Each of the above, except the last, make sense to me. For the last, I need to spend more time on it.

That said, reviewing the combination of them all is difficult.

Are all of these patches tied to https://github.com/mozilla/neqo/pull/2499? Can any of them merge into Neqo `main` in isolation?

martinthomson commented:
Sorry about the noise, but I'm not completely done yet.  It's still failing.

larseggert committed:
ci: No `mold` (#2857)

It takes longer to install the mold package then performing the link.

martinthomson committed:
Working!

martinthomson commented:
So, I just got everything working.  I'll need to go through and factor out all the chores and random improvements to logging that I added on the way through.  Then I will look at breaking out the PTO changes (which aren't super important, but will be essential for getting tests to pass) and the version negotiation stuff (which are important).  Some of the tests will probably not catch errors in the course of the changes, because we're not randomizing the initial packet numbers, but it's probably safest to stack these in the order of:

1. random cleanups, maybe in a few separate pull requests -- logging, making test connection parameters consistent, and maybe more -- whatever it takes to reduce noise for later patches
3. version negotiation fixes -- these are more important and it would be good to get those done quick-ish
4. PTO fixes
5. client initial-related changes

Because that now involves all the code from @larseggert as well, that's going to take a bit of work to basically invert the entire stack of changes, but that shouldn't be super difficult.

Isolating the version negotiation stuff will probably help with understanding it better.

martinthomson committed:
Add the early version upgrade back in

mxinden committed:
docs(http3): remove partially outdated docs (#2859)

martinthomson committed:
Enable randomization for the server

martinthomson committed:
Make the drop_handshake_packet_from_wrong_address robust

This test assumes that the server Initial is smaller than a single
packet.  That doesn't hold, even if we increase the size of that packet
by a single byte.  This test is a little more complex, but more robust.

martinthomson committed:
Disable randomization

martinthomson committed:
compatible_upgrade_large_initial robustness

This strips padding from all the packets in the handshake.
That makes all the accounting a lot simpler and robust.

This is not generic, but if I find a use for it in other tests (which
seems likely), it should be easy enough to factor it out.  The padding
test is a little fragile, but it accounts for the possibility of adding
SCONE at some point.

martinthomson committed:
Merge branch 'main' into non-determinism

martinthomson committed:
Merge branch 'strip-padding' into non-determinism

martinthomson committed:
Merge branch 'robust-wrong-hs-address' into non-determinism

martinthomson committed:
Make discarded_initial_keys more robust

Use split_datagram to make the dropping/dup counting more robust.
Rename some variables to make them more obvious.

martinthomson committed:
Fix import foul-up

martinthomson committed:
Merge branch 'robust-discarded-keys' into non-determinism

martinthomson committed:
Make only_server_initial more robust

This uses `split_datagram` again to ensure that our counting is accurate
in case we can't coalesce sometimes.

martinthomson committed:
Improve resending of Initial on undecryptable

The goal here was to make the test more robust, but it turns out that
the code could have been simpler as well.  That is, rather than testing
for a separate threshold, we can test instead for whether the saved
datagram store is full.  That was the same constant anyway.

The test is made more robust by having it disable MLKEM, which ensures
that there is just one Initial packet in each flight.  The undecryptable
packet is constructed manually (by rewriting an Initial).

martinthomson committed:
Make the coalesce_05rtt test robust

This was checking for dropped packets as if that made a significant
difference.  It does, because we want to make sure that no packet was
dropped because it couldn't be decrypted, but we need to ensure that the
packet padding wasn't masking real drops.  The way to fix that is to
strip padding where necessary.  Factors out the code from the previous
test.

martinthomson committed:
Completely rewrite dup_server_flight1

This was very old and it showed.  The new version is much more in line
with other tests we've written recently.  It also avoids a bunch of
unnecessary back and forth.  It covers all the cases properly as well.

martinthomson committed:
Test cleanup

martinthomson committed:
Move connection parameter defaults to the type

martinthomson committed:
Remove the first round of wrappers for ConnectionParameters

martinthomson committed:
Client doesn't need an ACK rate adjustment

martinthomson committed:
Revert all default parameters changes

martinthomson committed:
No need to rejig this now

martinthomson committed:
Remove randomization constraint, except for the ACK

martinthomson committed:
Add randomization to mitm_retry test

martinthomson committed:
Get the reasoning right here

martinthomson commented:
OK, I've spent a bunch of time cleaning this up today.

Most of that cleanup is in the form of making tests more robust.  Those tests now run against main and this branch, so rebasing here once all those pull requests merge should virtually eliminate most of this code.  That least two parts:
1. The version negotiation changes.
2. The randomization of the first packet number.

It will be a lot easier to split this change once the rest of the changes land, so I'm going to wait for those to work their way through the system.

larseggert commented:
We went with https://github.com/mozilla/neqo/pull/2885, so this is OBE.
